[WIP] Fix ht interface

matthiasbeyer opened the pull request:
Resubmit from #105 onto master

matthiasbeyer committed:
Fix: Return hash of element after insertion

matthiasbeyer committed:
The hashtable does not need to know the datasize

matthiasbeyer committed:
We need a compare element and the cfg as well when finding elements

matthiasbeyer committed:
We return the found element or NULL

matthiasbeyer committed:
Fix: ht_del() gets hash, cmp element, cfg and returns the element which was deleted

matthiasbeyer committed:
Fix: ht_insert() needs the cfg as well

matthiasbeyer committed:
Fix: The hash function gets void* and returns rs_hash

matthiasbeyer committed:
Fix: We should include the set header here as well

neithernut commented:
ACK.

matthiasbeyer commented:
waiting on CI.

matthiasbeyer commented:
CI failed because of stuff which should be fixed by #106 , waiting on this one.

matthiasbeyer committed:
Removed ht_find() implementation - reimplementation required

matthiasbeyer commented:
@neithernut Please review again, I removed `ht_find()` as the interface for this function completely changed. I will file an issue for reimplementing this one asap this PR is merged.

neithernut commented on src/libreset/ht.h:
```diff
@@ -95,13 +99,13 @@ ht_find(
+ * @return the hash of the inserted element
  */
-int
+rs_hash
```
Why return the hash? This way the function doesn't tell the user, whether the insertion was successful or not.

matthiasbeyer replied:
But the user has to know the hash of the inserted element so he can retreive it afterwards.

neithernut replied:
The hash is nowhere sufficient. The user may use a key of some soft, used by the hash function, to retrieve elements: the user may create a dummy object (a "comparable") and feed it to `ht_find`. That function will then search an object which will compare to the one provided and return that one to the user.
Also, the hash may be generated by the user anyways, since _he_ or _she_ provides the hash function in the first place.

matthiasbeyer replied:
You start to confuse me again. When inserting something, you get back a value to identify your object with. You have to pass this value to retreive the object from the library again.

So the hash is _relevant_ even if it is _not_ sufficient.

neithernut replied:
The user knows the internal structure of the object and is therefore better equipped to extract a key of some sort for identification of an object. In the end, the hash is nothing but a way to speed things up.

matthiasbeyer replied:
So you would like me to return a status and not the hash. Okay. So when the user inserts a value, her/his provided hash function is used for the hashtable, but the user will not get the hash, because  he can recreate it easily, am I right?

We should _definitely_ state in the docs that the hash is not supposed to be a hash in the known form.

neithernut replied:
What's a hash in the known form?
It is _completely_ common for hash table implementations to require the user to provide a hash function. See the C++ STL, or uthash, or ...
Hash functions may be provided for known standard types, but _no one_ would take the binary data of an object and compute a hash from that data, if the object may possibly not be a POD.

matthiasbeyer replied:
By "known form" I mean something like SHA or so, checksum hashes.

I never said that it is not common for libraries to ask the user for a hash function!

neithernut replied:
The underlying hash function may still be MD5 or SHA, but it is expected to be applied to only a part of the object and that the user will do so in the function he or she provides.

neithernut commented on src/libreset/ht.h:
```diff
@@ -111,10 +115,12 @@ ht_insert(
     struct ht* ht, //!< The hashtable object to delete from
-    rs_hash hash
+    rs_hash hash, //!< The hash of the element to delete
+    void* cmp, //!< Element to compare against
```
Maybe this function should use the same semantics `avl_del` uses now: take a predicate, instead of cmp and hash, and return the number of element removed. This way we can easily emulate the old behaviour while gaining new possibilities. However, this also slows things down because every bucket has to be accessed.
Maybe we should have two functions: one to remove elements based on a predicate and one based on a comparable. We could adapt the avl interface to have both versions, too. (open an issue then)

matthiasbeyer replied:
See #108 - I would like to keep the `_del` functions.

neithernut replied:
Agreed.

matthiasbeyer committed:
Fix: Return number of removed elements on ht_del()

neithernut commented:
Relates to issue #103.

matthiasbeyer committed:
Return status instead of hash

Due to the discussion in #107, we should return a status code here
instead of the hash.

neithernut commented on src/libreset/ht.h:
```diff
@@ -109,12 +113,14 @@ ht_insert(
+ * @return number of removed elements
  */
-struct element*
+unsigned int
```
Why not `int`?

matthiasbeyer replied:
because `int` _can_ be negative, and a counter is always positive. This way it is clear that a negative return value is not possible.

neithernut replied:
`ht_del()` _always_ deletes _at most_ one element. So returning whether the element was removed should suffice.

matthiasbeyer replied:
I thought we agreed on one function removing elements with similar hashes and an `_ndel` function removing also elements which have not the same hash, as discussed in #108

neithernut commented:
Also: please rebase. This may resolve the issue of the CI failing.

neithernut commented:
Ahhh, the CI fails because you renamed `hashf`.

matthiasbeyer committed:
Added missing include

neithernut commented:
ACK.
